Public/Src/Tools/SymbolDaemon/VsoSymbolClient.cs

```diff
@@ -0,0 +1,208 @@
+            new ArtifactHttpClientFactory(
+                credentials: GetCredentials(),
+                httpSendTimeout: m_config.HttpSendTimeout,
+                tracer: Tracer,
+                verifyConnectionCancellationToken: Token);
+
+        private Uri ServiceEndpoint => m_config.Service;
+
+        private string RequestName => m_config.Name;
+
+        private string RequestId
+        {
+            get
+            {
+                Contract.Requires(!string.IsNullOrEmpty(m_requestId));
+                return m_requestId;
+            }
+        }
+
+        /// <nodoc />
+        public VsoSymbolClient(ILogger logger, SymbolConfig config)
+        {
+            m_logger = logger;
+            m_config = config;
+            m_cancellationSource = new CancellationTokenSource();
+
+            m_logger.Info(I($"[{nameof(VsoSymbolClient)}] Using drop config: {JsonConvert.SerializeObject(m_config)}"));
+
+            m_symbolClient = CreateSymbolServiceClient();
+        }
+
+        private ISymbolServiceClient CreateSymbolServiceClient()
+        {
+            var client = new SymbolServiceClient(
+                ServiceEndpoint,
+                GetFactory(),
+                Tracer,
+                new SymbolServiceClientTelemetry(Tracer, ServiceEndpoint, enable: m_config.EnableTelemetry));
+
+            return client;
+        }
+
+        /// <summary>
+        /// Queries the endpoint for the RequestId. 
+        /// This method should be called only after the request has been created, otherwise, it will throw an exception.
+        /// </summary>
+        /// <remarks>
+        /// On workers, m_requestId won't be initialized, so we need to query the server for the right value.
+        /// </remarks>
+        private async Task EnsureRequestIdInitalizedAsync()
+        {
+            if (string.IsNullOrEmpty(m_requestId))
+            {
+                var result = await m_symbolClient.GetRequestByNameAsync(RequestName, Token);
+                m_requestId = result.Id;
+            }
+        }
+
+        /// <summary>
+        /// Creates a symbol request.
+        /// </summary>
+        public async Task<Request> CreateAsync(CancellationToken token)
+        {
+            var result = await m_symbolClient.CreateRequestAsync(RequestName, token);
+
+            m_requestId = result.Id;
+
+            return result;
+        }
+
+        /// <inheritdoc />
+        public Task<Request> CreateAsync() => CreateAsync(Token);
+
+        /// <inheritdoc />
+        public async Task<AddDebugEntryResult> AddFileAsync(SymbolFile symbolFile)
+        {
+            Contract.Requires(symbolFile.IsIndexed, "File has not been indexed.");
+            Contract.Requires(symbolFile.DebugEntries.Count > 0, "File contains no symbol data.");
+
+            await EnsureRequestIdInitalizedAsync();
+
+            var result = await m_symbolClient.CreateRequestDebugEntriesAsync(

```

Comment:
>CreateRequestDebugEntriesAsync [](start = 46, length = 30)

this method can be called for multiple debug entries for multiple files, right? All those debug entries do not need to represent one file, i guess. that's wny, instead of calling this method for one file, we can use it for batching, right?

Reply:
correct, I'll add 'optimizations' in another PR

---
In reply to: [314219841](https://github.com/microsoft/BuildXL/pull/727#discussion_r314219841) [](ancestors = 314219841)